transaction/std/src/signed_contingent_input_builder.rs

```diff
@@ -0,0 +1,2272 @@
+//// Copyright (c) 2022 The MobileCoin Foundation
+
+//! A builder object for signed contingent inputs (see MCIP #31)
+//! This plays a similar role to the transaction builder.
+
+use crate::{ChangeDestination, InputCredentials, MemoBuilder, TxBuilderError};
+use core::cmp::min;
+use mc_account_keys::PublicAddress;
+use mc_fog_report_validation::FogPubkeyResolver;
+use mc_transaction_core::{
+    ring_signature::{GeneratorCache, OutputSecret, RingMLSAG, Scalar, SignableInputRing},
+    tx::{TxIn, TxOut, TxOutConfirmationNumber},
+    Amount, BlockVersion, InputRules, MemoContext, MemoPayload, NewMemoError,
+    SignedContingentInput, UnmaskedAmount,
+};
+use rand_core::{CryptoRng, RngCore};
+
+/// Helper utility for creating signed contingent inputs with required outputs,
+/// and attaching fog hint and memos as appropriate.
+///
+/// This is generic over FogPubkeyResolver because there are several reasonable
+/// implementations of that.
+///
+/// This is generic over MemoBuilder to allow injecting a policy for how to
+/// use the memos in the TxOuts.
+#[derive(Debug)]
+pub struct SignedContingentInputBuilder<FPR: FogPubkeyResolver> {
+    /// The block version that we are targeting for this transaction
```

Comment:
```suggestion
    /// The block version that we are targeting for this input
```

This is making me wonder about whether a contingent input should include a block version in it...
Might be useful to have that information in case there are some contingent inputs floating around, and we switch to a newer block version that makes the old ones incompatible. Right now I speculate that this would manifest as weird signature errors or such, depending on where the incompatibility comes from.

Reply:
I feel like the contingent input should try to be block version agnostic. It seems like an undesireable behaviour for a contract to be invalidated by a backend change to the block version.

Reply:
that is a really good point, thank you

Reply:
> I feel like the contingent input should try to be block version agnostic. It seems like an undesireable behaviour for a contract to be invalidated by a backend change to the block version.

Yeah for sure, but it may be unavoidable. For instance, in block version 2 vs. block version 3 we changed the hash that was being signed by the MLSAGs. There's no way a contingent input could be compatible across a change like that.

We may be able to make the transaction builder smart though, like, if no changes that would impact singed contignent inputs happen between block version 3 and block version 4, then maybe the transaction builder can still accept signed contingent inputs at block version 3 even when it's building for block version 4.

I feel like it can't hurt anything to put a block version in the signed contingent input payload -- the `Tx` proto still doesn't itself have a block version, and the transaction builder can make a decision if it thinks it will work or not, before dropping the block version value from the input.